### PR TITLE
fix(expr): fix a critical bug in case expression

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Sometimes the CASE expression fails to short circuit, causing the last matching value to be returned instead of the first one. 

```diff
 dev=> create table t(x boolean, a int, y boolean, b int);
 CREATE_TABLE
 dev=> insert into t values (true, 1, true, 2);
 INSERT 0 1
 dev=> select *, case when x then a when y then b end from t;
  x | a | y | b | case 
 ---+---+---+---+------
- t | 1 | t | 2 |    2
+ t | 1 | t | 2 |    1
```

The reason is that some expressions **ignore the visibility and return non-null values for invisible rows.** Examples include literal expression and simd-optimized expressions such as `=`. The CASE expression doesn't mask invisible rows in the condition result, so later values may override earlier ones.

This bug seems to have been there from ~~day 1~~ v0.1.15 (Jan 2023). It sounds very horrible. 🤯

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [x] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

### CODE REVIEW

File: src/expr/impl/src/scalar/case.rs [hunk 1]
```diff
@@ -65,8 +61,10 @@ impl Expression for CaseExpression {
         let when_len = self.when_clauses.len();
         let mut result_array = Vec::with_capacity(when_len + 1);
         for (when_idx, WhenClause { when, then }) in self.when_clauses.iter().enumerate() {
-            let calc_then_vis = when.eval(&input).await?.as_bool().to_bitmap();
             let input_vis = input.visibility().clone();
+            // note that evaluated result from when clause may contain bits that are not visible,
+            // so we need to mask it with input visibility.
+            let calc_then_vis = when.eval(&input).await?.as_bool().to_bitmap() & &input_vis;
```
Comment (wangrunji0408) [inline src/expr/impl/src/scalar/case.rs:67]: The core fix.

**Review by stdrc — APPROVED** [review]
LGTM! Thx for the quick fix!

**Comment by wangrunji0408** [comment]
I confirmed that this bug was introduced during v0.1.14 to v0.1.15, by #6581.

**Comment by xxchan** [comment]
Is this considered a breaking change? 🥵 https://github.com/risingwavelabs/rfcs/pull/24#issuecomment-1488470861 @BugenZhao

**Comment by BugenZhao** [comment]
> Is this considered a breaking change? 🥵 https://github.com/risingwavelabs/rfcs/pull/24#issuecomment-1488470861 @BugenZhao 
> 
> 

Strictly speaking, yes. 😕 We need to introduce a new function type with the fixed implementation, while keeping the buggy one as "backend only". But it's also okay for me to simply assume that no one persisted the result prior to this PR. 🤣

**Pull request merged**